### PR TITLE
Add register template

### LINKED ISSUE #86: Add register template
**Expected Behavior**
A user will be able to register to the app. 

**Current Behavior**
no registration page.

### PR DESCRIPTION
**Pull request by IdanZimi**
### Desired Outcome

Registration page

### Implemented Changes

Added register template, view and tests. User can register.

### Connected Issue/Story

closes #86 

### Dependencies
- [x] This PR does not depend on other PR's 

### Definition of Done
1. Add register.html template
2. Add path /register at meetballs/urls
3. Add registerUser view at meet_balls_app/views
4. Add tests for registration page at meet_balls_app/tests

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes
  
  
image

### CODE REVIEW

File: meet_balls_app/views.py [hunk 1]
```diff
@@ -27,3 +31,63 @@ def loginUser(request):
...
+            return render(request, 'meet_balls_app/register.html',
+                          {'ball_games': BallGame.choices})
+        else:
+            try:
+                player = Player.create(
+                    username, password, birthdate, favorite_ball_game)
+                user = player.user
+                login(request, user)
+                messages.error(request, '')
+                return render(request, 'meet_balls_app/home.html',
```
Comment (OhadSeg) [inline meet_balls_app/views.py:64]: @IdanZimi I think it should be `redirect` rather `render`, and maybe it should be better to redirect the user to the login page, where he should log in, and not directly log him in, when he registrates.

File: meet_balls_app/views.py [hunk 2]
```diff
@@ -27,3 +31,63 @@ def loginUser(request):
...
+                error_messages = e.messages[0].split("\n")
+                for error in error_messages:
+                    messages.error(request, error)
+                    return render(request, "/meet_balls_app/register.html")
+    else:
+        messages.error(request, '')
+        return render(request, 'meet_balls_app/register.html', {'ball_games': BallGame.choices})
+
+
+def validate_register_form(data):
```
[the remark this answers is not in the document]
Comment (IdanZimi) [reply meet_balls_app/views.py:76]: No, the element we are using in the register form is "select" so he can only choose out of finite set of choices.
`<select class="form-select" name="favorite_ball_game" required>
                        {% for ball_game in ball_games %}
                        <option value="{{ ball_game.0 }}">{{ ball_game.1 }}</option>
                        {% endfor %}
                    </select>`

File: meet_balls_app/templates/meet_balls_app/register.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                            {% endfor %}
+                        </ul>
+                    </div>
+                </div>
+            </div>
+            {% endif %}
+        </div>
+    </div>
+</div>
+{% endblock %}
```
[the remark this answers is not in the document]
Comment (IdanZimi) [reply meet_balls_app/templates/meet_balls_app/register.html]: Fixed

File: meet_balls_app/views.py (hunk 1 above)
Comment (IdanZimi) [reply meet_balls_app/views.py]: Fixed. Now It redirects to the login page after successful registration. @OhadSeg

File: meet_balls_app/tests.py [hunk 4]
```diff
@@ -64,6 +73,82 @@ def test_successful_logout(self, client, player):
...
         logout_url = reverse('logoutUser')
         response = client.post(logout_url)
+        assert response.status_code == 302
+        assert response.url == reverse('home')
+        assert response.wsgi_request.user.is_authenticated is False
+
+
+@pytest.mark.django_db
+class TestRegister:
+    def test_registerUser_valid(self, client):
```
Comment (Yuval-Vino) [inline meet_balls_app/tests.py:83]: Shouldn't we test whether the player has been created and now exists?

**Comment by OhadSeg** [comment]
@IdanZimi With the latest changes I think the PR looks good, precise, and on-point.

File: meet_balls_app/tests.py [hunk 5]
```diff
@@ -64,6 +73,82 @@ def test_successful_logout(self, client, player):
...
+        assert response.url == reverse('loginUser')
+
+    @pytest.mark.parametrize('username, password, confirm_password, birthdate, favorite_ball_game', [
+        ('', 'passwrod', 'password', '1990-01-01', 'Soccer'),
+        ('testuser', '', 'password', '1990-01-01', 'Soccer'),
+        ('testuser', 'passwrod', '', '1990-01-01', 'Soccer'),
+        ('testuser', 'passwrod', 'password', '', 'Soccer'),
+        ('testuser', 'passwrod', 'password', '1990-01-01', '')
+    ])
+    def test_registerUser_invalid(self, client, username, password, confirm_password, birthdate, favorite_ball_game):
```
Comment (Yuval-Vino) [inline meet_balls_app/tests.py:104]: How about considering a more descriptive name that clearly signifies a test scenario without a required field? This could help avoid confusion, especially since you already have a test named `test_registerUser_passwords_dont_match`, which could also be perceived as `test_registerUser_invalid`.

File: meet_balls_app/tests.py (hunk 4 above)
Comment (IdanZimi) [reply meet_balls_app/tests.py]: Agree. Fixed.

**Review by Danielsio — COMMENTED** [review]
Hi Great Work So far !!
Could you please add fields for first name and last name of user ?

File: meet_balls_app/tests.py [hunk 6]
```diff
@@ -64,6 +76,101 @@ def test_successful_logout(self, client, player):
...
+            'password': 'testpassword',
+            'confirm_password': 'testpassword',
+            'birthdate': '1990-01-01',
+            'favorite_ball_game': 'Soccer'
+        }
+
+        register_url = reverse('registerUser')
+        response = client.post(register_url, data)
+        assert Player.objects.filter(user__username='testuser').exists()
+        assert Player.objects.filter(user__username='testuser').count() == 1
```
Comment (MaayanMashhadi) [inline meet_balls_app/tests.py:98]: If I understand correctly, you already check if the 'testuser' exists.
Is it necessary checking the count?  Or is it necessary to check the exists?

**Review by MaayanMashhadi — COMMENTED** [review]
Good work, the validation looks good for me!

File: meet_balls_app/templates/meet_balls_app/register.html [hunk 7]
```diff
@@ -0,0 +1,53 @@
+{% extends 'base.html' %}
+
+{% block content %}
+<div class="container">
+    <div class="row justify-content-center">
+        <div class="col-md-6 col-lg-4">
+            <h2 class="text-center mb-4">Register</h2>
+            <form method="post">
+                {% csrf_token %}
+                <div class="mb-3">
```
Comment (amitCohen2) [inline meet_balls_app/templates/meet_balls_app/register.html:10]: Please change these classes names to meaningful names,
It will help us to work with it in CSS files

File: meet_balls_app/tests.py [hunk 8]
```diff
@@ -64,6 +76,101 @@ def test_successful_logout(self, client, player):
...
+        assert response.status_code == 200
+        assert 'Passwords do not match.' in response.content.decode()
+        assert response.templates[0].name == 'meet_balls_app/register.html'
+
+    def test_unsuccessful_register_birthdate_in_future(self, client):
+        future_date = timezone.now() + datetime.timedelta(days=1)
+        data = {
+            'username': 'testuser',
+            'password': 'password',
+            'confirm_password': 'password1',
```
Comment (amitCohen2) [inline meet_balls_app/tests.py:158]: I think the `'confirm_password': 'password1,'` should be here `'confirm_password': 'password',`

**Comment by IdanZimi** [comment]
> Hi Great Work So far !! Could you please add fields for first name and last name of user ?

Where were going to save these fields? we only have username as part of the model. @Danielsio

File: meet_balls_app/tests.py (hunk 6 above)
Comment (IdanZimi) [reply meet_balls_app/tests.py:98]: I was thinking maybe it is necessary to check that only 1 user was created and no  more. I agree It might be redundant.
What do you think? @yftacherzog @Yuval-Vino

File: meet_balls_app/templates/meet_balls_app/register.html (hunk 7 above)
Comment (IdanZimi) [reply meet_balls_app/templates/meet_balls_app/register.html:10]: Assuming these names are bootstrap names. Is it possible to change them? Isn't it ruin the current design?

File: meet_balls_app/tests.py (hunk 8 above)
Comment (IdanZimi) [reply meet_balls_app/tests.py:158]: Agree. fixed.

File: meet_balls_app/tests.py (hunk 6 above)
Comment (Yuval-Vino) [reply meet_balls_app/tests.py:98]: I believe it's redundant, particularly since it's not possible to create a user with an existing username (which has been tested in Player tests).

Comment (IdanZimi) [reply meet_balls_app/tests.py]: fixed

File: meet_balls_app/views.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SchSeba) [reply meet_balls_app/views.py:72]: That is right but nothing blocks me from sending a POST request directly via something like postman (https://www.postman.com/) :)

validation should always be done in both frond and back end

File: meet_balls_app/templates/meet_balls_app/register.html (hunk 7 above)
Comment (amitCohen2) [reply meet_balls_app/templates/meet_balls_app/register.html:10]: Good point. Use it like that for not have some bootstrap issues

File: meet_balls_app/views.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (IdanZimi) [reply meet_balls_app/views.py:72]: Ok. I added validation at the `validate_register_form` method. 
In addition, I added a test for this case `test_unsuccessful_register_invalid_ball_game @SchSeba `

**Comment by amitKaplansky** [comment]
this pr doesn't depend on style pr?

**Comment by amitKaplansky** [comment]
To ensure visibility of players on events, we need to add the fields "First Name" and "Last Name" to the registration form. If these fields are left empty, the player's information will not be displayed on the events they participate in.

File: meet_balls_app/views.py
Comment (amitKaplansky) [inline meet_balls_app/views.py:1]: Is there any password validation, such as a minimum number of characters or any specific requirements?

Comment (IdanZimi) [reply meet_balls_app/views.py:1]: Only that password matches the confirm password field. I don't think we need to handle this at the moment. 
maybe in the future.

**Comment by IdanZimi** [comment]
> To ensure visibility of players on events, we need to add the fields "First Name" and "Last Name" to the registration form. If these fields are left empty, the player's information will not be displayed on the events they participate in.

Why we cant display the username of players who intend to participate at the event. @amitKaplansky

**Review by SchSeba — APPROVED** [review]
LGTM

**Review by Yuval-Vino — APPROVED** [review]
LGTM!

**Comment by amitKaplansky** [comment]
> > To ensure visibility of players on events, we need to add the fields "First Name" and "Last Name" to the registration form. If these fields are left empty, the player's information will not be displayed on the events they participate in.
> 
> Why we cant display the username of players who intend to participate at the event. @amitKaplansky

I saw you added first and last name, that is just as we did it, it made more sense to be the name of the player.

File: meet_balls_app/tests.py [hunk 9]
```diff
@@ -64,6 +73,82 @@ def test_successful_logout(self, client, player):
         client.post(login_url, {'username': username, 'password': password})
         logout_url = reverse('logoutUser')
         response = client.post(logout_url)
+        assert response.status_code == 302
+        assert response.url == reverse('home')
+        assert response.wsgi_request.user.is_authenticated is False
+
+
+@pytest.mark.django_db
+class TestRegister:
```
[the remark this answers is not in the document]
Comment (amitKaplansky) [reply meet_balls_app/tests.py:85]: Did you add a test to validate first and last name are not empty? that the name is only letters and not any other characters?

**Comment by yftacherzog** [comment]
@IdanZimi can we merge this one?

**Comment by IdanZimi** [comment]
> @IdanZimi can we merge this one?

Yes @yftacherzog
